### ISSUE TITLE
Integrate lfortran with an LLM

### ISSUE THREAD

**Issue by Beliavsky** [lfortran/lfortran#1594]
Integrating a Fortran compiler with a large language model, as I suggested at https://groups.google.com/g/comp.lang.fortran/c/GVASMUKv_Ww , would be interesting. That post is copied below.

I have asked ChatGPT to write Python and Fortran codes to solve some problems, and often it produces working code, but sometimes it produces Fortran code that does not compile. It would be great to have a ChatGPT wrapper with access to a free compiler such as gfortran and more broadly to all the compilers in gcc. The workflow would be

(1) Ask the ChatGPT wrapper a programming question.
(2a) If ChatGPT answers without producing code, transmit the answer to the user.
(2b) If it does produce code in Fortran or C++ etc., try to compile with gfortran, g++ etc.
(2bi) If the code compiles, transmit the answer with code to the user.
(2bii) If the code does not compile, send the error message to ChatGPT, which hopefully it can fix, after which the answer with code is transmitted to the user.

The basic idea is to improve ChatGPT so that it does not send the user code that does not compile. One could extend this by having the ChatGPT wrapper not just compile but run the codes it generates and test if it satisfies the user's requirements. For example, the user could say "write a Fortran function called 'sorted' in module 'sorted_mod' that sorts integers, and test it with a program

use sorted_mod
implicit none
if (all(sorted([5,4,2]) == [2, 4, 5]) then
print*,"success"
else
print*,"failure"
end if
end"

There is a project https://github.com/mherrmann/fix using ChatGPT to fix C++ programs.

**Comment by certik** [comment]
Yes. Even better would be to get this working with model that you can run locally. One of the best seem https://vicuna.lmsys.org/, it seems it is still not quite there, but it seems it might be borderline useful. I am hopeful that in a year or two we'll see some solid LLM that can run locally and is useful. This interface should be designed to work with any LLM. I also think even more robust would be to train the LLM to output LFortran's ASR, in the Clojure form (S-expressions) that we use. It would eliminate large number of issues (all syntax and surface level semantics). The best would be to design such intermediate representation that every program that it produces is valid and compiles. Then if LLVM produces that, it will always compile and run, so it would remove the "failure" degrees of freedom, potentially making it easier for LLVM to produce higher quality code.
